runtime/op/scope.go

```diff
@@ -0,0 +1,60 @@
+package op
+
+import (
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/zbuf"
+)
+
+type enterScope struct {
+	parent zbuf.Puller
+	vars   []zed.Value
+}
+
+func NewEnterScope(parent zbuf.Puller, vars []zed.Value) zbuf.Puller {
+	if len(vars) == 0 {
+		return parent
+	}
+	return &enterScope{parent, vars}
+}
+
+func (s *enterScope) Pull(done bool) (zbuf.Batch, error) {
+	batch, err := s.parent.Pull(done)
+	if batch != nil {
+		vars := append(batch.Vars(), s.vars...)
```

Comment:
Batches can be accessed concurrently, so appending to batch.Vars() is racy. Clone it first.
```suggestion
		vars := append(slices.Clone(batch.Vars()), s.vars...)
```

Reply:
Good catch.